# Project 3 -- Taylor Saunders

**TA Help:** 
na
    
**Collaboration:** 
na

## Question 1

### Explore the files in the provided data set directory. Find out how many years are included in the data set. Briefly describe the contents of the files.

In [1]:
import pandas as pd

In [2]:
from pathlib import Path

In [3]:
import os

In [4]:
Path("/anvil/projects/tdm/data/noaa/1750.csv")

PosixPath('/anvil/projects/tdm/data/noaa/1750.csv')

In [5]:
len(os.listdir('/anvil/projects/tdm/data/noaa/'))

266

There are 266 years in the dataset. There are two sample files for 2020, but the rest of the files are individual years ranging from 1763 to 2023.

### Create a list named myfiles, to hold Path objects from 1880.csv to 1883.csv in the data set folder using list comprehension.

In [6]:
myfiles=[ Path(f'/anvil/projects/tdm/data/noaa/{year}.csv') for year in range (1880,1884) ]
print(myfiles)

[PosixPath('/anvil/projects/tdm/data/noaa/1880.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1881.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1882.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1883.csv')]


## Question 2

### Calculate how many records are in the file 1880.csv. (Each line is one record.)

In [7]:
file = Path(f'/anvil/projects/tdm/data/noaa/1880.csv')
with open(file,"r") as f:
    mycount = 0
    for line in f:
        mycount += 1
print(f'There are {mycount} records in the file called {file}')

There are 370779 records in the file called /anvil/projects/tdm/data/noaa/1880.csv


### Calculate how many records there are (altogether) in the 4 files from 1880.csv to 1883.csv. Use the list myfiles that you created in Question 1. Your output should give the total number of records altogether

In [8]:
for year in range (1880, 1884):
    file = Path(f'/anvil/projects/tdm/data/noaa/{year}.csv')
    with open(file,"r") as f:
        mycount = 0
        for line in f:
            mycount += 1
    print(f'There are {mycount} records in the file called {file}')

There are 370779 records in the file called /anvil/projects/tdm/data/noaa/1880.csv
There are 443270 records in the file called /anvil/projects/tdm/data/noaa/1881.csv
There are 514083 records in the file called /anvil/projects/tdm/data/noaa/1882.csv
There are 578730 records in the file called /anvil/projects/tdm/data/noaa/1883.csv


## Question 3

### Run the following statement, to read in the first file from the list myfiles into a DataFrame using myDF = pd.read_csv(myfiles[0]). Display the column names for myDF. Look at the head and tail of myDF. Do you see anything unexpected?

In [9]:
myfiles[0]

PosixPath('/anvil/projects/tdm/data/noaa/1880.csv')

In [10]:
myDF = pd.read_csv(myfiles[1])

In [11]:
myDF.head()

AGE00135039  18810101  PRCP   30 Unnamed: 4 Unnamed: 5  E  Unnamed: 7
0  AGE00147705  18810101  PRCP  479        NaN        NaN  E         NaN
1  AGE00147708  18810101  PRCP  125        NaN        NaN  E         NaN
2  AGE00147709  18810101  PRCP  125        NaN        NaN  E         NaN
3  AGE00147711  18810101  PRCP    0        NaN        NaN  E         NaN
4  AGE00147712  18810101  PRCP  138        NaN        NaN  E         NaN

There are 3 "unnamed" columns. By using read_csv, it uses the first row as the header.

### Please modify your work from Question 3a, to correct the problem that you found. What are the column names now?

In [12]:
myDF = pd.read_csv(myfiles[0], header = None)

In [13]:
myDF.head()

0         1     2   3    4    5  6   7
0  AGE00147705  18800101  PRCP   0  NaN  NaN  E NaN
1  AGE00147708  18800101  PRCP  36  NaN  NaN  E NaN
2  AGE00147709  18800101  PRCP  36  NaN  NaN  E NaN
3  AGE00147712  18800101  PRCP   0  NaN  NaN  E NaN
4  AGE00147713  18800101  TMAX  52  NaN  NaN  E NaN

The column names are now numerical, "0, 1, 2, 3, 4, 5, 6, 7"

### Now let us add these 7 column names: id, date, element_code, value, mflag, qflag, sflag, and obstime to the data frame.

In [14]:
pd.read_csv(myfiles[0],names = ["id","date","element_code","value","mflag","qflag","sflag","obstime"])

id      date element_code  value mflag qflag sflag  obstime
0       AGE00147705  18800101         PRCP      0   NaN   NaN     E      NaN
1       AGE00147708  18800101         PRCP     36   NaN   NaN     E      NaN
2       AGE00147709  18800101         PRCP     36   NaN   NaN     E      NaN
3       AGE00147712  18800101         PRCP      0   NaN   NaN     E      NaN
4       AGE00147713  18800101         TMAX     52   NaN   NaN     E      NaN
...             ...       ...          ...    ...   ...   ...   ...      ...
370774  USW00093852  18801231         TMAX     39   NaN   NaN     X      NaN
370775  USW00093852  18801231         TMIN    -17   NaN   NaN     X      NaN
370776  USW00094728  18801231         TMAX   -122   NaN   NaN     0      NaN
370777  USW00094728  18801231         TMIN   -194   NaN   NaN     0      NaN
370778  USW00094728  18801231         PRCP      0   NaN   NaN     0      NaN

[370779 rows x 8 columns]

### Make a list called mydataframes (of length 4) that contains 4 data frames, one for each year, from 1880.csv to 1883.csv. Starting with the sample code (above) for reading in the first file, modify our example, so that you have a "for" loop that reads in all 4 files. Test your work with a for loop that displays the column names of each of the four data frames in mydataframes.

In [15]:
mydataframes = []
for i in range(0,4):
    mydataframes.append(pd.read_csv(myfiles[0],names = ["id","date","element_code","value","mflag","qflag","sflag","obstime"]))

In [16]:
mydataframes[1].shape

(370779, 8)

In [17]:
for i in range(0,4):
    print(mydataframes[i].shape)

(370779, 8)
(370779, 8)
(370779, 8)
(370779, 8)


In [18]:
for i in range(0,4):
    print(mydataframes[i].columns)

Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')
Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')
Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')
Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')


## Question 4

### Let’s look at the column element_code. Use a loop to solve the following questions for all 4 DataFrames:

### Print out the (unique) elements of the column element_code (i.e., show each element just one time).

In [19]:
mydataframes[i]['element_code'].unique()

array(['PRCP', 'TMAX', 'TMIN', 'SNOW', 'SNWD', 'TAVG', 'MDPR', 'DATN',
       'DATX', 'MDTN', 'MDTX', 'DAPR', 'DWPR', 'MDSF', 'WT01', 'TOBS',
       'WT04', 'WT05'], dtype=object)

In [20]:
for myDF in mydataframes:
    print(myDF['element_code'].unique)

<bound method Series.unique of 0         PRCP
1         PRCP
2         PRCP
3         PRCP
4         TMAX
          ... 
370774    TMAX
370775    TMIN
370776    TMAX
370777    TMIN
370778    PRCP
Name: element_code, Length: 370779, dtype: object>
<bound method Series.unique of 0         PRCP
1         PRCP
2         PRCP
3         PRCP
4         TMAX
          ... 
370774    TMAX
370775    TMIN
370776    TMAX
370777    TMIN
370778    PRCP
Name: element_code, Length: 370779, dtype: object>
<bound method Series.unique of 0         PRCP
1         PRCP
2         PRCP
3         PRCP
4         TMAX
          ... 
370774    TMAX
370775    TMIN
370776    TMAX
370777    TMIN
370778    PRCP
Name: element_code, Length: 370779, dtype: object>
<bound method Series.unique of 0         PRCP
1         PRCP
2         PRCP
3         PRCP
4         TMAX
          ... 
370774    TMAX
370775    TMIN
370776    TMAX
370777    TMIN
370778    PRCP
Name: element_code, Length: 370779, dtype: object>


In [21]:
for myDF in mydataframes:
    print(len(myDF['element_code'].unique()))

18
18
18
18


### Find the number of times that SNOW occurs in the element_code column.

In [22]:
for myDF in mydataframes:
    print(myDF['element_code'].value_counts()['SNOW'])

33700
33700
33700
33700


## Question 5

### Try to run the following 2 programs, to find the number of times that SNOW occurs in the element_code column, from the year 1880 to year 1883. Explain your understanding of chunksize.

In [23]:
myfiles=[]
for year in range (1880, 1884):
    file= Path(f'/anvil/projects/tdm/data/noaa/{year}.csv')
    myfiles.append(file)

In [28]:
count = 0
for file in myfiles:
    for myDF in pd.read_csv(file,names=["id","date","element_code","value","mflag","qflag","sflag","obstime"],chunksize =10000):
        count += len(myDF[myDF['element_code'] == 'SNOW'])
print(count)

164216


In [27]:
count = 0
for file in myfiles:
    for myDF in pd.read_csv(file,names=["id","date","element_code","value","mflag","qflag","sflag","obstime"],chunksize =10000):
        for index, row in myDF.iterrows():
            if row['element_code'] == 'SNOW':
                count += 1

print(count)

164216


Chunksize helps run time when working with large datasets; "chunksize" is the number of rows that will be read when a program is ran, regardless of the datasets size. 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.